sdk/go/pulumi/properties.go

```diff
@@ -32,290 +40,145 @@ type Output struct {
-			}
-		}()
-	} else {
-		out.s.sync <- &valueOrError{value: v, known: known}
-	}
-}
+	state uint32 // one of output{Pending,Resolved,Rejected}
 
-// reject will reject the output.  It is not exported, because we want to control the capabilities tightly, such
-// that anybody who happens to have an Output is not allowed to reject it; only those who created it can.
-func (out *Output) reject(err error) {
-	out.s.sync <- &valueOrError{err: err}
-}
+	value interface{} // the value of this output if it is resolved.
+	err   error       // the error associated with this output if it is rejected.
+	known bool        // true if this output's value is known.
 
-// Apply transforms the data of the output property using the applier func.  The result remains an output property,
-// and accumulates all implicated dependencies, so that resources can be properly tracked using a DAG.  This function
-// does not block awaiting the value; instead, it spawns a Goroutine that will await its availability.
-func (out *Output) Apply(applier func(v interface{}) (interface{}, error)) *Output {
-	result, resolve, reject := NewOutput(out.Deps())
-	go func() {
-		for {
-			v, known, err := out.Value()
-			if err != nil {
-				reject(err)
-				break
-			} else {
-				if known {
-					// If we have a known value, run the applier to transform it.
-					u, err := applier(v)
-					if err != nil {
-						reject(err)
-						break
-					} else {
-						// Now that we've transformed the value, it's possible we have another output.  If so, pluck it
-						// out and go around to await it until we hit a real value.  Note that we are not capturing the
-						// resources of this inner output, intentionally, as the output returned should be related to
-						// this output already.
-						if newout, ok := v.(*Output); ok {
-							out = newout
-						} else {
-							resolve(u, true)
-							break
-						}
-					}
-				} else {
-					// If the value isn't known, skip the apply function.
-					resolve(nil, false)
-					break
-				}
-			}
-		}
-	}()
-	return result
+	deps []Resource // the dependencies associated with this output property.
 }
 
-// Deps returns the dependencies for this output property.
-func (out *Output) Deps() []Resource { return out.s.deps }
+func (o *outputState) fulfill(value interface{}, known bool, err error) {
+	o.mutex.Lock()
 
-// Value retrieves the underlying value for this output property.
-func (out *Output) Value() (interface{}, bool, error) {
-	// If neither error nor value are available, first await the channel.  Only one Goroutine will make it through this
-	// and is responsible for closing the channel, to signal to other awaiters that it's safe to read the values.
-	if out.s.voe == nil {
-		if voe := <-out.s.sync; voe != nil {
-			out.s.voe = voe   // first time through, publish the value.
-			close(out.s.sync) // and close the channel to signal to others that the memozied value is available.
-		}
+	if o.state != outputPending {
+		o.mutex.Unlock()
```

Comment:
what do you think about doing a top level `o.mutex.Unlock()` to cover all once and future branches?

Reply:
That should be fine, yes. An earlier iteration of this method was running callbacks and needed to be careful about holding the lock. I'll make the change.